calico/felix/actor.py

```diff
@@ -450,11 +455,17 @@ class TrackedAsyncResult(AsyncResult):
     """
     def __init__(self, tag):
         super(TrackedAsyncResult, self).__init__()
-        self.__ref = ExceptionTrackingRef(self, _reap_ref)
-        self.__ref.tag = tag
-
-    def set_msg(self, msg):
-        self.__ref.msg = msg
+        # Avoid keeping a reference to the week ref directly; look it up
+        # when needed.  Also, be careful not to attach any debugging
+        # information to the ref that could produce a reference cycle.  The
+        # tag should be something simple like a string or tuple.
+        tr = ExceptionTrackingRef(self, _reap_ref)
+        tr.tag = tag
+        self.__ref_idx = tr.idx
+
+    @property
+    def __ref(self):
+        return _refs[self.__ref_idx]
```

Comment:
No, it can't, the ref is removed from the dict when the TrackedAsyncResult is GCed so there's no way to hit this line after the ref is removed from the dict.

Basically, this code is intended to call _reap_ref when the TrackedAsyncResult gets GCed.  It's almost equivalent to just using a __del__ method on the TrackedAsyncResult but, because it uses a weakref to get the callback, it doesn't mess up garbage collection.

The annoying part is that the weakref callback doesn't get called if the weakref dies before the TrackedAsyncResult does so we have to stash it in a strongly-referenced dict and then remove it after the TrackedAsyncResult gets reaped.  (We can't stash the weak ref on the TrackedAsyncResult because an objects fields get GCed before it does!)